ludwig/benchmarking/reporting.py

```diff
@@ -33,3 +39,92 @@ def create_metrics_report(experiment_name: str) -> Tuple[Dict[str, Any], str]:
     merged_file_path = os.path.join(os.getcwd(), experiment_name, "metrics_report", "{}.json".format("full_report"))
     save_json(merged_file_path, full_report)
     return full_report, merged_file_path
+
+
+def initialize_stats_dict(function_events):
+    """Initialize dictionary which stores resource usage information per tagged code block."""
+    info = dict()
+    for event_name in [evt.name for evt in function_events if "ludwig" in evt.name]:
+        info[event_name] = {"code_block_tag": event_name}
+        info[event_name]["runs"] = []
+    return info
+
+
+def get_memory_details(event):
+    """Get device name and number of bytes (de)allocated during an event."""
+    if event.device_type() in [DeviceType.CPU, DeviceType.MKLDNN, DeviceType.IDEEP]:
+        return "cpu", event.nbytes()
+    elif event.device_type() in [DeviceType.CUDA, DeviceType.HIP]:
+        return f"cuda_{event.device_index()}", event.nbytes()
+
+
+def get_devices_usage(kineto_event, mem_records_acc, run_usage_info):
+    """Get CPU and CUDA memory usage for and event."""
+    memory_so_far = defaultdict(int)
+    memory_lists = defaultdict(list)
+    for mem_record in mem_records_acc.in_interval(
+        kineto_event.start_us(), kineto_event.start_us() + kineto_event.duration_us()
+    ):
+        device, nbytes = get_memory_details(mem_record[0])
+        memory_so_far[device] += nbytes
+        memory_lists[device].append(memory_so_far[device])
+    for device in memory_lists:
+        memory_lists[device].append(0)  # just in case we have an empty list
+        run_usage_info[f"average_{device}_memory_usage"] = mean(memory_lists[device])
+        run_usage_info[f"max_{device}_memory_usage"] = max(memory_lists[device])
+        run_usage_info[f"average_{device}_memory_usage_str"] = _format_memory(mean(memory_lists[device]))
+        run_usage_info[f"max_{device}_memory_usage_str"] = _format_memory(max(memory_lists[device]))
+    return run_usage_info
+
+
+def get_device_timing(function_event, run_usage_info):
```

Comment:
The docstring says "run durations". Maybe `get_device_run_durations()`, as this is more clear than "timings"